File: graphview/src/main/java/de/blox/graphview/GraphAdapter.java
package de.blox.graphview;

import android.support.annotation.NonNull;
import android.view.View;
import android.widget.Adapter;

import de.blox.graphview.tree.BuchheimWalkerAlgorithm;

/**
 *
 */
public interface GraphAdapter<VH> extends Adapter, NodeObserver {

    void notifySizeChanged();

    /**
     * Returns the currently set algorithm. It uses the {@link BuchheimWalkerAlgorithm} as default,
     * if no algorithm is previously set.
     *
     * @return
     */
    Algorithm getAlgorithm();

    /**
     * Set an algorithm, which is used for laying out the tree.
     *
     * @param algorithm the algorithm to use for laying out the tree
     */
    void setAlgorithm(@NonNull Algorithm algorithm);


    /**
     * Set a graph.
     * @param graph
     */
    void setGraph(@NonNull Graph graph);

    /**
     * Returns the node at a given {code position}.
     *
     * @param position
     * @return
     */
    Node getNode(int position);

    /**
     * Returns the screen position from the node at {code position}
     *
     * @param position
     * @return
     */
    Vector getScreenPosition(int position);

    @NonNull
    VH onCreateViewHolder(View view);

    void onBindViewHolder(VH viewHolder, Object data, int position);

    Graph getGraph();
}

File: graphview/src/main/java/de/blox/graphview/GraphAdapter.kt
package de.blox.graphview

import android.database.DataSetObservable
import android.database.DataSetObserver
import android.view.View
import android.view.ViewGroup
import android.widget.Adapter
import androidx.annotation.NonNull

abstract class GraphAdapter<VH : GraphView.ViewHolder>(var graph: Graph) : Adapter {

    private val dataSetObservable = DataSetObservable()
    private var graphViewObserver: DataSetObserver? = null

    fun notifyDataSetChanged() {
        synchronized(this) {
            graphViewObserver?.onChanged()
        }
        dataSetObservable.notifyChanged()
    }

    override fun registerDataSetObserver(observer: DataSetObserver) {
        dataSetObservable.registerObserver(observer)
    }

    override fun unregisterDataSetObserver(observer: DataSetObserver) {
        dataSetObservable.unregisterObserver(observer)
    }

    open fun setGraphViewObserver(observer: DataSetObserver?) {
        synchronized(this) { graphViewObserver = observer }
    }

    @NonNull
    abstract fun onCreateViewHolder(@NonNull parent: ViewGroup, viewType: Int): VH

    abstract fun onBindViewHolder(@NonNull viewHolder: VH, data: Any, position: Int)

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val view: View
        val viewHolder: VH

        if (convertView == null) {
            viewHolder = onCreateViewHolder(parent, getItemViewType(position))
            view = viewHolder.itemView
            view.tag = viewHolder
        } else {
            viewHolder = convertView.tag as VH
            view = viewHolder.itemView
        }

        onBindViewHolder(viewHolder, getItem(position), position)

        return view
    }

    override fun getItemId(position: Int): Long = NO_ID

    override fun hasStableIds(): Boolean = false

    override fun getItemViewType(position: Int): Int = 0

    override fun getViewTypeCount(): Int = 0

    companion object {
        const val NO_ID: Long = -1
    }
}
